### PR TITLE
Ingest

### PR DESCRIPTION
**Pull request by sreyashinag**
A Gobblin pipeline used to extract metadata from files stored in HDFS and write to Ground.

### CODE REVIEW

File: ground-core/src/main/java/edu/berkeley/ground/api/models/Node.java [hunk 1]
```diff
@@ -27,7 +27,7 @@
     private String name;
 
     @JsonCreator
-    protected Node(
+    public Node(
```
Comment (vsreekanti) [inline ground-core/src/main/java/edu/berkeley/ground/api/models/Node.java:30]: Can we make all of the API models public while we're at it please?

File: ground-ingest/pom.xml [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+    <name>Ground Ingest</name>
+    
+    <properties>
+        <gobblin.version>0.8.0</gobblin.version>
+    </properties>
+    
+    <parent>
+        <groupId>edu.berkeley</groupId>
+        <artifactId>ground</artifactId>
+        <version>0.1-SNAPSHOT</version>
```
Comment (vsreekanti) [inline ground-ingest/pom.xml]: This version should be inherited from the parent POM for now.

File: ground-ingest/src/main/java/edu/berkeley/ground/ingest/FileMetadataSource.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+import gobblin.source.extractor.filebased.FileBasedHelperException;
+import gobblin.source.extractor.filebased.FileBasedSource;
+import gobblin.source.extractor.hadoop.HadoopFsHelper;
+import gobblin.util.HadoopUtils;
+
+
+/**
+ * A Source that extracts metadata from a file system.
+ */
+public class FileMetadataSource extends FileBasedSource<Schema, GenericRecord> {
```
Comment (vsreekanti) [inline ground-ingest/src/main/java/edu/berkeley/ground/ingest/FileMetadataSource.java]: Is it canonical to put a Gobblin source not-in-Gobblin?

File: ground-ingest/src/main/java/edu/berkeley/ground/ingest/GroundDataWriterBuilder.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+import gobblin.writer.DataWriterBuilder;
+
+
+public class GroundDataWriterBuilder extends DataWriterBuilder<Schema, GenericRecord> {
+
+  @Override
+  public DataWriter<GenericRecord> build() throws IOException {
+    return new GroundWriter<>();
+  }
+
```
Comment (vsreekanti) [inline ground-ingest/src/main/java/edu/berkeley/ground/ingest/GroundDataWriterBuilder.java]: whitespace

File: ground-ingest/src/main/java/edu/berkeley/ground/ingest/GroundWriter.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+public class GroundWriter<D> implements DataWriter<GenericRecord>  {
+  
+    public void write(GenericRecord record) throws IOException {
+    
+      //extracting the metadata from the GenericRecord and storing it in the field tags      
+      Map<String, Tag> tags =  new HashMap<String, Tag>();
+      GroundType typeString = GroundType.STRING;
+      GroundType typeInt = GroundType.INTEGER;
+      Tag timeCreated = new Tag("id5", "timeCreated", record.get("timeCreated"), typeInt);
```
Comment (vsreekanti) [inline ground-ingest/src/main/java/edu/berkeley/ground/ingest/GroundWriter.java]: You should just be able to pass in an empty string?

File: ground-ingest/src/main/java/edu/berkeley/ground/ingest/GroundWriter.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+  } 
+    
+    //method to create a Node given the name
+    public void node(String name, Map<String, Tag> tags) throws JsonGenerationException, JsonMappingException, IOException {
+     
+      Node node = new Node("id", name);
+      ObjectMapper mapper = new ObjectMapper();
+      String jsonString = mapper.writeValueAsString(node);
+      HttpClient client = new HttpClient();
+      String uri = "http://localhost:9090/nodes/"+ name;
```
Comment (vsreekanti) [inline ground-ingest/src/main/java/edu/berkeley/ground/ingest/GroundWriter.java]: This shouldn't be hardcoded.

**Review by vsreekanti — COMMENTED** [review]
Like we talked about last time, all of the ground-ingest stuff should be within the `plugins/` directory. There's also still some loose formatting (lots of extra whitespace, mostly). Beyond that, more or less looks good.  Left a couple questions & comments in line, but we're pretty close!

File: ground-ingest/src/main/java/edu/berkeley/ground/ingest/FileMetadataSource.java (hunk 3 above)
Comment (shirshanka) [reply ground-ingest/src/main/java/edu/berkeley/ground/ingest/FileMetadataSource.java]: Yes this should be fine. Gobblin expects that users will extend / implement their own sources.

File: ground-ingest/src/main/java/edu/berkeley/ground/ingest/GroundWriter.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+   
+  } 
+    
+    //method to create a Node given the name
+    public void node(String name, Map<String, Tag> tags) throws JsonGenerationException, JsonMappingException, IOException {
+     
+      Node node = new Node("id", name);
+      ObjectMapper mapper = new ObjectMapper();
+      String jsonString = mapper.writeValueAsString(node);
+      HttpClient client = new HttpClient();
```
Comment (shirshanka) [inline ground-ingest/src/main/java/edu/berkeley/ground/ingest/GroundWriter.java]: HttpClient should be member variable inside the writer, so you're not creating a new client per node.

File: ground-ingest/src/main/resources/run_scheduler.sh [hunk 8]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+CLASSPATH=/Users/sreyashinag/Documents/workspace/sreyashi-ground-fork/ground-ingest/target/ground-ingest-0.1-SNAPSHOT-jar-with-dependencies.jar
+GOBBLIN_FWDIR="$(cd `dirname $0`/..; pwd)"
+FWDIR_CONF=$GOBBLIN_FWDIR/conf
+GOBBLIN_LOG_DIR="$GOBBLIN_FWDIR/logs"
+DEFAULT_CONFIG_FILE=$FWDIR_CONF/gobblin-standalone.properties
+GOBBLIN_CUSTOM_CONFIG_FILE=$CUSTOM_CONFIG_FILE
+GOBBLIN_JVM_FLAGS=$JVM_FLAGS
+$JAVA_HOME/bin/java -Xmx2g -Xms1g -XX:+UseConcMarkSweepGC -XX:+UseParNewGC -XX:+PrintGCDetails -XX:+PrintGCDateStamps -XX:+PrintTenuringDistribution -XX:+UseCompressedOops -XX:+HeapDumpOnOutOfMemoryError -XX:HeapDumpPath=$GOBBLIN_LOG_DIR/ -Xloggc:$GOBBLIN_LOG_DIR/gobblin-gc.log -Dgobblin.logs.dir=$GOBBLIN_LOG_DIR -Dlog4j.configuration=file://$FWDIR_CONF/log4j-standalone.xml -cp $CLASSPATH -Dorg.quartz.properties=$FWDIR_CONF/quartz.properties $GOBBLIN_JVM_FLAGS gobblin.scheduler.SchedulerDaemon $DEFAULT_CONFIG_FILE $GOBBLIN_CUSTOM_CONFIG_FILE
```
Comment (shirshanka) [inline ground-ingest/src/main/resources/run_scheduler.sh]: This file should be deleted. The real run_scheduler.sh seems to be in ground-ingest/src/main/scripts/run_scheduler.sh

File: plugins/ground-ingest/src/main/java/edu/berkeley/ground/ingest/GroundWriter.java [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+      tags.put("timeCreated", timeCreated);
+      tags.put("length", length);
+      tags.put("modificationTime", modificationTime);
+      tags.put("owner", owner);
+        
+      //extracting the name of the file to create the Node
+      String name = record.get("name").toString();
+      node(name, tags);
+   
+  }
```
Comment (vsreekanti) [inline plugins/ground-ingest/src/main/java/edu/berkeley/ground/ingest/GroundWriter.java:54]: formatting? please fix

Comment (vsreekanti) [reply plugins/ground-ingest/src/main/java/edu/berkeley/ground/ingest/GroundWriter.java:54]: formatting is weird throughout this file

File: pom.xml [hunk 10]
```diff
@@ -9,6 +9,7 @@
 
     <modules>
         <module>ground-core</module>
+        <module>ground-ingest</module>
```
Comment (vsreekanti) [inline pom.xml]: This should be `plugins/`.

**Pull request merged**